File: laboratorios/lab03/l-java/src/main/java/com/lab/hilos/trapecio/Trapecio.java
package com.lab.hilos.trapecio;

import com.ezylang.evalex.Expression;
import com.ezylang.evalex.data.EvaluationValue;
import java.math.BigDecimal;
import java.math.MathContext;
import java.util.Scanner;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.CountDownLatch;

public class Trapecio {
    private static double areaTotal = 0.0;
    private static final Object areaMutex = new Object();
    private static String funcionMatematica;

    public static double evaluarFuncion(String expresion, double valorX) {
        try {
            // Reemplazar x por el valor
            String exprConValor = expresion.replace("x", String.valueOf(valorX));
            
            Expression expr = new Expression(exprConValor);
            EvaluationValue result = expr.evaluate();
            
            return result.getNumberValue().doubleValue();
        } catch (Exception e) {
            System.out.println("Error al compilar la funcion: '" + expresion + "'");
            return 0.0;
        }
    }

    public static double calcularAreaTrapecio(double baseMenor, double baseMayor, double altura) {
        return (baseMenor + baseMayor) * altura / 2.0;
    }

    public static void calcularTrapecio(double limiteA, double limiteB, int indiceHilo, int totalHilos, CountDownLatch latch) {
        try {
            double altura = (limiteB - limiteA) / totalHilos;
            double x0 = limiteA + indiceHilo * altura;
            double x1 = limiteA + (indiceHilo + 1) * altura;

            double fX0 = evaluarFuncion(funcionMatematica, x0);
            double fX1 = evaluarFuncion(funcionMatematica, x1);

            double areaTrapecio = calcularAreaTrapecio(fX0, fX1, altura);

            synchronized (areaMutex) {
                areaTotal += areaTrapecio;
            }

            System.out.printf("=> Hilo %d: [%f, %f] -> Area = %f%n", indiceHilo, x0, x1, areaTrapecio);
        } finally {
            latch.countDown();
        }
    }

    public static void main(String[] args) {
        System.out.println("---------- CALCULADORA DE INTEGRALES CON HILOS NORMALES ----------");
        System.out.println();

        System.out.println("Ingrese la función f(x) a integrar:");
        System.out.println("   Ejemplos:");
        System.out.println("   - 2*x^2 + 3*x + 0.5");
        System.out.println("   - SIN(x)");
        System.out.println("   - COS(x) + x^2");
        System.out.println("   - EXP(x)");
        System.out.println("   - LOG(x)");
        System.out.println("   - SQRT(x)");
        System.out.println();
        System.out.print("f(x) = ");

        Scanner scanner = new Scanner(System.in);
        funcionMatematica = scanner.nextLine().trim();

        if (funcionMatematica.isEmpty()) {
            System.out.println("Error: No se ingreso ninguna funcion");
            return;
        }

        System.out.println("Funcion ingresada: '" + funcionMatematica + "'");

        // validar funcion
        try {
            double test1 = evaluarFuncion(funcionMatematica, 1.0);
            double test2 = evaluarFuncion(funcionMatematica, 2.0);
            if (test1 != 0.0 || test2 != 0.0) {
                System.out.println("Funcion valida");
            }
        } catch (Exception e) {
            System.out.println("Error: La funcion no es valida");
            System.exit(1);
        }

        double limiteInferior, limiteSuperior;
        int decimales, numeroHilos;

        System.out.print("Limite inferior (a): ");
        limiteInferior = scanner.nextDouble();
        System.out.print("Limite superior (b): ");
        limiteSuperior = scanner.nextDouble();
        System.out.print("Numero de hilos (trapecios): ");
        numeroHilos = scanner.nextInt();
        System.out.print("Decimales en el resultado: ");
        decimales = scanner.nextInt();
        System.out.println();

        // validaciones
        if (limiteSuperior <= limiteInferior) {
            System.out.println("Error: El limite superior debe ser mayor al inferior");
            return;
        }

        if (numeroHilos <= 0) {
            System.out.println("Error: El numero de hilos debe ser positivo");
            return;
        }

        if (numeroHilos > 1000) {
            System.out.println("Error: Muchos hilos pueden afectar el rendimiento");
        }

        if (decimales < 0 || decimales > 10) {
            System.out.println("Error: El numero de decimales debe estar entre 0 y 10");
            return;
        }

        System.out.println("Iniciando calculo...");

        CountDownLatch latch = new CountDownLatch(numeroHilos);

        for (int i = 0; i < numeroHilos; i++) {
            final int indice = i;
            new Thread(() -> calcularTrapecio(limiteInferior, limiteSuperior, indice, numeroHilos, latch)).start();
        }

        try {
            latch.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        System.out.println();
        System.out.println("RESULTADO FINAL:");
        System.out.printf("   Area = %.*f%n", decimales, areaTotal);
        System.out.println();

        scanner.close();
    }
}

File: laboratorios/lab03/l-go/trapecio.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strings"
    "sync"
    "github.com/Pramod-Devireddy/go-exprtk"
)

var (
    areaTotal   float64
    areaMutex   sync.Mutex
    funcionMat  string
)

func evaluarFuncion(expresion string, valorX float64) float64 {
    exprtkObj := exprtk.NewExprtk()
    defer exprtkObj.Delete()

    exprtkObj.SetExpression(expresion)
    exprtkObj.AddDoubleVariable("x")

    err := exprtkObj.CompileExpression()
    if err != nil {
        fmt.Printf("Error al compilar la funcion: '%s' - %s\n", expresion, err.Error())
        return 0.0
    }

    exprtkObj.SetDoubleVariableValue("x", valorX)
    return exprtkObj.GetEvaluatedValue()
}

func calcularAreaTrapecio(baseMenor, baseMayor, altura float64) float64 {
    return (baseMenor + baseMayor) * altura / 2.0
}

func calcularTrapecio(limiteA, limiteB float64, indiceHilo, totalHilos int, wg *sync.WaitGroup) {
    defer wg.Done()
    
    altura := (limiteB - limiteA) / float64(totalHilos)
    x0 := limiteA + float64(indiceHilo)*altura
    x1 := limiteA + float64(indiceHilo+1)*altura
    
    fX0 := evaluarFuncion(funcionMat, x0)
    fX1 := evaluarFuncion(funcionMat, x1)
    
    areaTrapecio := calcularAreaTrapecio(fX0, fX1, altura)
    
    areaMutex.Lock()
    areaTotal += areaTrapecio
    areaMutex.Unlock()
    
    fmt.Printf("=> Hilo %d: [%f, %f] -> Area = %f\n", indiceHilo, x0, x1, areaTrapecio)
}

func main() {
    fmt.Println("---------- CALCULADORA DE INTEGRALES CON HILOS NORMALES ----------")
    fmt.Println()
    
    fmt.Println("Ingrese la función f(x) a integrar:")
    fmt.Println("   Ejemplos:")
    fmt.Println("   - 2*x^2 + 3*x + 0.5")
    fmt.Println("   - sin(x)")
    fmt.Println("   - cos(x) + x^2")
    fmt.Println("   - exp(x)")
    fmt.Println("   - log(x)")
    fmt.Println("   - sqrt(x)")
    fmt.Println()
    fmt.Print("f(x) = ")
    
    scanner := bufio.NewScanner(os.Stdin)
    scanner.Scan()
    funcionMat = strings.TrimSpace(scanner.Text())
    
    if funcionMat == "" {
        fmt.Println("Error: No se ingreso ninguna funcion")
        return
    }
    
    fmt.Printf("Funcion ingresada: '%s'\n", funcionMat)
    
    // Validar funcion
    test1 := evaluarFuncion(funcionMat, 1.0)
    test2 := evaluarFuncion(funcionMat, 2.0)
    fmt.Printf("f(1) = %f, f(2) = %f\n", test1, test2)
    fmt.Println("Funcion valida")
    
    var limiteInferior, limiteSuperior float64
    var decimales, numeroHilos int
    
    fmt.Print("Limite inferior (a): ")
    fmt.Scanf("%f", &limiteInferior)
    fmt.Print("Limite superior (b): ")
    fmt.Scanf("%f", &limiteSuperior)
    fmt.Print("Numero de hilos (trapecios): ")
    fmt.Scanf("%d", &numeroHilos)
    fmt.Print("Decimales en el resultado: ")
    fmt.Scanf("%d", &decimales)
    fmt.Println()
    
    // Validaciones
    if limiteSuperior <= limiteInferior {
        fmt.Println("Error: El limite superior debe ser mayor al inferior")
        return
    }
    
    if numeroHilos <= 0 {
        fmt.Println("Error: El numero de hilos debe ser positivo")
        return
    }
    
    if numeroHilos > 1000 {
        fmt.Println("Error: Muchos hilos pueden afectar el rendimiento")
    }
    
    if decimales < 0 || decimales > 10 {
        fmt.Println("Error: El numero de decimales debe estar entre 0 y 10")
        return
    }
    
    fmt.Println("Iniciando calculo...")
    
    var wg sync.WaitGroup
    
    for i := 0; i < numeroHilos; i++ {
        wg.Add(1)
        go calcularTrapecio(limiteInferior, limiteSuperior, i, numeroHilos, &wg)
    }
    
    wg.Wait()
    
    fmt.Println()
    fmt.Println("RESULTADO FINAL:")
    fmt.Printf("   Area = %.*f\n", decimales, areaTotal)
    fmt.Println()
}